src/CoreWCF.WebHttp/src/CoreWCF/FreezableCollection.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace CoreWCF
+{
+    internal class FreezableCollection<T> : Collection<T>, ICollection<T>
+    {
```

Comment:
I don't think we need to add this class. The location where it's used exposes a `Collection<T>`. You could make FastPathInfo._candidates and FastPathInfo._relativePathSegments not readonly and use `Collection<T>`, then in FastPathInfo.Freeze, replace them with `ReadOnlyCollection<T>` wrapping the existing ones. This would make them readonly and still avoids any copies as ReadOnlyCollection passes through read operations to the original collection.

Reply:
That seems like it would work for the FastPathInfo stuff, but I don't know if it would work for UriTemplatesCollection without a fair bit of rework.

Reply:
I'm still stuck on this one. If I change UriTemplatesCollection to a Collection<KeyValuePair<UriTemplate, object>> how do I change it into a read only collection? I can't cast from a ReadOnlyCollection<KeyValuePair<UriTemplate, object>>> to a UriTemplatesCollection